--- generated_code/csrc/includes/ds_kernel_utils.h	2023-06-08 22:43:13.713967768 -0700
+++ manual_fixed/csrc/includes/ds_kernel_utils.h	2023-06-08 22:43:29.110099794 -0700
@@ -30,7 +30,7 @@
 
 #if DPCT_COMPATIBILITY_TEMP >= 530
 #define HALF_PRECISION_AVAILABLE = 1
-#define PTX_AVAILABLE
+/* #define PTX_AVAILABLE */
 #endif  // __CUDA_ARCH__ >= 530
 
 #if DPCT_COMPATIBILITY_TEMP >= 800
@@ -40,6 +40,9 @@
 
 #endif  //__HIP_PLATFORM_HCC__
 
+using uint4 = sycl::vec<uint, 4>;
+using uint2 = sycl::vec<uint, 2>;
+
 inline int next_pow2(const int val)
 {
     int rounded_val = val - 1;
